gammapy/utils/fitting/sampling.py

```diff
@@ -0,0 +1,162 @@
+    for factor, par in zip(pars, dataset.parameters.free_parameters):
+        par.factor = factor
+
+    lnprob_priors = lnprior(dataset)
+
+    # dataset.likelihood returns Cash statistics values
+    # emcee will maximisise the LogLikelihood so we need -dataset.likelihood
+    total_lnprob = -dataset.likelihood() + lnprob_priors
+
+    return total_lnprob
+
+
+def run_mcmc(dataset, nwalkers=8, nrun=1000, threads=1):
+    """
+    Run the MCMC sampler.
+
+    Parameters
+    ----------
+    dataset : `MapDataset`
+        A gammapy dataset object. This contains the observed counts cube,
+        the exposure cube, the psf cube, and the sky model and model.
+        Each free parameter in the sky model is considered as parameter for the MCMC.
+    nwalkers: int
+        Required integer number of walkers to use in ensemble.
+        Minimum is 2*nparam+2, but more than that is usually better.
+        Must be even to use MPI mode.
+    nrun: int
+        Number of steps for walkers. Typically at least a few hundreds (but depends on dimensionality).
+        Low nrun (<100?) will underestimate the errors.
+        Samples that would populate the distribution are nrun*nwalkers.
+        This step can be ~seen as the error estimation step.
+    """
+    dataset.parameters.autoscale()  # Autoscale parameters
+    pars = model_to_par(dataset)  # get a tuple of parameters from dataset
+    ndim = len(pars)
+
+    # Initialize walkers in a ball of relative size 0.5% in all dimensions if the
+    # parameters have been fit, or to 10% otherwise
+    spread = 0.5 / 100
+    p0var = np.array([spread * pp for pp in pars])
+    p0 = emcee.utils.sample_ball(pars, p0var, nwalkers)
+
+    labels = []
+    for par in dataset.parameters.free_parameters:
+        labels.append(par.name)
+        if (par.min is np.nan) and (par.max is np.nan):
+            print(
+                    "Warning: no priors have been set for parameter %s\n The MCMC will likely not work !"
+                    % (par.name)
+            )
+
+    print("List of free parameters:", labels)
+    print("{} walkers will run for {} steps".format(nwalkers, nrun))
+    print("Parameters init value for 1st walker:", p0[0])
+    sampler = emcee.EnsembleSampler(
+        nwalkers, ndim, lnprob, args=[dataset], threads=threads
+    )
+
+    for idx, result in enumerate(sampler.sample(p0, iterations=nrun)):
+        if (idx + 1) % 100 == 0:
+            print("{0:5.0%}".format(idx / nrun))
+
+    return sampler
+
+def plot_trace(sampler, dataset):
+    """
+    Plot the trace of walkers for every steps
+    """
+    labels = []
+    for par in dataset.parameters.free_parameters:
+        labels.append(par.name)
+
+    fig, ax = plt.subplots(len(labels), sharex=True)
+    for i in range(len(ax)):
+        ax[i].plot(sampler.chain[:, :, i].T, "-k", alpha=0.2)
+        ax[i].set_ylabel(labels[i])
+    plt.xlabel("Nrun")
+
+
+def plot_corner(sampler, dataset, nburn=0):
```

Comment:
Document `dataset` in `Parameters`.

Maybe consider taking parameters instead of dataset. So users would pass `dataset.parameters.free_parameters` when calling often, but if they have 10 free parameters, they might just pass the subset of 3 parameters of interest at the moment for a plot?

Reply:
I like the fact that the user doesn't have to handle which param is free or not and just provide a high level object.
But maybe we could just pass the parameters: `dataset.parameters`.
but the call will be a bit longer `plot_corner(sampler, dataset.parameters)` instead of `plot_corner(sampler, dataset)`  and not as much plug&play.

The same is valid for `plot_trace(sampler, dataset)`.

Reply:
@facero - We don't have a clear design in the modelling & fitting code yet in which interfaces `Parameters` or `Model` or `Dataset` objects are passed. So put as you prefer for now.